app/services/hts_service/reports/hts_report_builder.rb

```diff
@@ -38,115 +38,94 @@ def his_patients_rev
-              data[patient_id] = patient_data
-            end
-        
-            patient_data["observations"] << patient_obs
-          end
-        
-          return process_patient_data(data, indicators)
-    end
-        
-    def process_patient_data(data, indicators)
-          patient_data = []
+      def his_patients_revs(indicators) 
+    
+                columns = ActiveRecord::Base.connection.columns('obs').map(&:name) 
+                sql_query = <<~SQL
+                SELECT
+                person.gender,
+                person.birthdate,
+                encounter.encounter_datetime,
+                obs.*
+              FROM patient
+              INNER JOIN person ON person.person_id = patient.patient_id
+              INNER JOIN encounter ON encounter.patient_id = patient.patient_id
+              INNER JOIN program ON program.program_id = encounter.program_id
+              INNER JOIN obs ON obs.person_id = patient.patient_id
+              WHERE
+                patient.voided = 0
+                AND encounter.encounter_datetime BETWEEN '#{@start_date}' AND '#{@end_date}'
+                AND encounter.encounter_type = (SELECT encounter_type_id FROM encounter_type WHERE name = 'Testing')
+                AND program.program_id = (SELECT program_id FROM program WHERE name = 'HTC PROGRAM')
+                AND obs.voided = 0;              
+             SQL
+    
+         results = ActiveRecord::Base.connection.select_all(sql_query)
+         return process_patient_data(results, indicators)
+            
+      end
+            
+      def process_patient_data(results, indicators)  
+    
+        data = []
+    
+        grouped_obs = results.group_by { |obs| obs['person_id'] }
         
-          data.each do |_patient_id, patient_observation|
-            observations = patient_observation["observations"]
-            patient_obs = {}
+        grouped_obs.each_with_index do |row, index|
+          patient_id = row[0]
+          observations = row[1]    
+          data << { "person_id" => patient_id,"gender"=> observations[0]["gender"],"birthdate"=>observations[0]['birthdate'],"encounter_datetime"=>observations[0]["encounter_datetime"] }    
+          index_new = data.index(data.last)     
         
-            indicators.each do |indicator|
-              if indicator[:concept_id].is_a?(Integer)
-                desired_observation = observations.find { |obs| obs["concept_id"] == indicator[:concept_id].to_s }
+          indicators.each do |indicator|
+            if indicator[:concept_id].is_a?(Integer)
+              desired_observation = observations.find { |obs| obs["concept_id"] == indicator[:concept_id] }
+              if desired_observation.present?              
+                val = indicator[:value]
+                name = indicator[:name]
```

Comment:
Just curious how does the front-end handle response of indicators with indexes and those without?

For example line 86 and line 97 are different